File: MobileApp/lib/models/workout_schedule.dart
import 'package:workout_app/models/schedule_day.dart';

class WorkoutSchedule {
  int id;
  String name;
  String? description;
  DateTime startDate;
  DateTime? endDate;
  bool isActive;
  List<ScheduleDay> days;

  WorkoutSchedule({
    required this.id,
    required this.name,
    this.description,
    required this.startDate,
    this.endDate,
    required this.isActive,
    required this.days,
  });
}

extension WorkoutScheduleCopyWith on WorkoutSchedule {
  WorkoutSchedule copyWith({
    int? id,
    String? name,
    String? description,
    DateTime? startDate,
    DateTime? endDate,
    bool? isActive,
    List<ScheduleDay>? days,
  }) {
    return WorkoutSchedule(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      isActive: isActive ?? this.isActive,
      days: days ?? this.days,
    );
  }
}

File: API/WorkoutAPI/Models/WorkoutSchedule.cs
namespace WorkoutAPI.Models
{
    public class WorkoutSchedule
    {
        public int ID { get; set; }
        public required string Name { get; set; }
        public string Description { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime? EndDate { get; set; }
        public List<ScheduleDay> Days { get; set; }
        public WorkoutSchedule() { }
        public WorkoutSchedule(int id, string name, string description, DateTime startDate, DateTime? endDate, List<ScheduleDay> days)
        {
            ID = id;
            Name = name;
            Description = description;
            StartDate = startDate;
            EndDate = endDate;
            Days = days;
        }
        public bool IsActive()
        {
            return EndDate == null || EndDate.Value.Date <= DateTime.Today;
        }

        public ScheduleDay GetActiveDay()
        {
            int daysSinceStart = (DateTime.Today - StartDate.Date).Days;
            if (daysSinceStart < 0)
                daysSinceStart = 0;
            return Days[daysSinceStart % Days.Count];
        }
    }
}
